spartan/src/cli/commands/replica.rs

```diff
@@ -0,0 +1,155 @@
+        {
+            Replication::Replica(config) => TcpListener::bind(config.host)
+                .await
+                .map_err(|e| ReplicaCommandError::SocketError(e)),
+            _ => Err(ReplicaCommandError::ReplicaConfigNotFound),
+        }?;
+
+        loop {
+            match socket.accept().await {
+                Ok((socket, _)) => {
+                    ReplicaSocket::new(&manager, socket)
+                        .exchange(accept_connection)
+                        .await?
+                }
+                Err(e) => Err(ReplicaCommandError::SocketError(e))?,
+            }
+        }
+    }
+}
+
+struct ReplicaSocket<'a> {
+    manager: &'a Manager<'a>,
+    socket: Framed<TcpStream, BytesCodec>,
+}
+
+impl<'a> ReplicaSocket<'a> {
+    fn new(manager: &'a Manager<'a>, socket: TcpStream) -> Self {
+        ReplicaSocket {
+            manager,
+            socket: BytesCodec::new().framed(socket),
+        }
+    }
+
+    async fn exchange<F, Fut>(&mut self, f: F) -> Result<(), ReplicaCommandError>
+    where
+        F: Fn(PrimaryRequest<'static>, &'a Manager<'a>) -> Fut,
+        Fut: Future<Output = ReplicaRequest>,
+    {
+        loop {
+            let buf = match self.socket.next().await {
+                Some(r) => r.map_err(|e| ReplicaCommandError::SocketError(e))?,
+                None => Err(ReplicaCommandError::EmptySocket)?,
+            };
+
+            let request = f(
+                deserialize(&buf).map_err(|e| ReplicaCommandError::SerializationError(e))?,
+                self.manager,
+            )
+            .await;
+
+            self.socket
+                .send(
+                    serialize(&request)
+                        .map_err(|e| ReplicaCommandError::SerializationError(e))?
+                        .into(),
+                )
+                .await
+                .map_err(|e| ReplicaCommandError::SocketError(e))?;
+
+            self.socket
+                .flush()
+                .await
+                .map_err(|e| ReplicaCommandError::SocketError(e))?;
+        }
+    }
+}
+
+async fn accept_connection(
+    request: PrimaryRequest<'static>,
+    manager: &Manager<'_>,
+) -> ReplicaRequest {
+    match request {
+        PrimaryRequest::Ping => ReplicaRequest::Pong,
+        PrimaryRequest::AskIndex => {
+            let mut indexes = Vec::with_capacity(manager.config.queues.len());
+
+            for (name, db) in manager.node.iter() {
+                let index = db
+                    .lock()
+                    .await
+                    .get_storage()
+                    .as_mut()
+                    .expect("No database present")
+                    .get_replica()
+                    .get_index();
+
+                indexes.push((name.to_string().into_boxed_str(), index));
+            }
+
+            ReplicaRequest::RecvIndex(indexes.into_boxed_slice())
+        }
+        PrimaryRequest::SendRange(queue, range) => ReplicaRequest::RecvRange,
```

Comment:
Applying incoming actions and setting confirmed_index to index of last element in range goes here